src/modules/0.0.22/aws_s3/mappers/bucket.ts

```diff
@@ -38,74 +75,131 @@ export class BucketMapper extends MapperBase<Bucket> {
     }
   }
   updateBucketPolicy = crudBuilder2<S3, 'putBucketPolicy'>('putBucketPolicy', input => input);
+
+  async createBucketPolicy(client: S3, bucket: Bucket, ctx: Context) {
+    const input: PutBucketPolicyCommandInput = {
+      Bucket: bucket.name,
+      Policy: JSON.stringify(bucket.policyDocument ?? {}),
+    };
+    await this.updateBucketPolicy(client, input);
+
+    // requery the created policy from AWS
+    const inputGet: GetBucketPolicyCommandInput = {
+      Bucket: bucket.name,
+    };
+    const bucketPolicy = await this.getBucketPolicy(client, inputGet);
+    if (bucketPolicy && bucketPolicy.Policy) {
+      bucket.policyDocument = JSON.parse(bucketPolicy.Policy);
+    } else {
+      bucket.policyDocument = undefined;
+    }
+    await this.module.bucket.db.update(bucket, ctx);
+
+    return bucket.policyDocument;
+  }
+
   cloud = new Crud2<Bucket>({
     // TODO: There are lots of useful permission controls on create to be added to this model
     create: async (es: Bucket[], ctx: Context) => {
-      const client = (await ctx.getAwsClient()) as AWS;
       const out = [];
+      const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
+      let client;
+
       for (const e of es) {
+        // if the bucket name exists in another region we cannot create it, throw error
+        for (const region of enabledRegions) {
+          client = (await ctx.getAwsClient(region)) as AWS;
+          try {
+            await this.headBucket(client.s3Client, e.name);
+            throw new Error('Cannot create the bucket, it already exists in another region');
+          } catch (_) {
+            // it is ok, we can create bucket
+          }
+        }
+
+        // we can create the bucket
+        client = (await ctx.getAwsClient(e.region)) as AWS;
         await this.createBucket(client.s3Client, e.name);
+
         out.push(e);
       }
       return out;
     },
     read: async (ctx: Context, id?: string) => {
-      const client = (await ctx.getAwsClient()) as AWS;
-
-      const allBuckets = await this.getBuckets(client.s3Client);
+      const client = (await ctx.getAwsClient(await ctx.getDefaultRegion())) as AWS;
+      const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
       const out: Bucket[] = [];
-      const buckets: BucketAWS[] = allBuckets.filter(b => !id || b.Name === id).filter(b => !!b.Name);
+      let rawBuckets: BucketAWS[] = [];
 
-      for (const index in buckets) {
-        if (buckets.hasOwnProperty(index)) {
-          const bucket = buckets[index];
+      if (!!id) {
+        const { bucketName, region } = this.idFields(id);
+        if (enabledRegions.includes(region)) {
+          const allBuckets = await this.getBuckets(client.s3Client);
+          rawBuckets = allBuckets.filter(b => !bucketName || b.Name === bucketName).filter(b => !!b.Name);
+        }
+      } else {
```

Comment:
if this method is being called with an `id`, we have to return just one element and not a list of them, so I think we should `.find` the `bucketName` received from the id, get its policy, map it and return it.